red_queen/games/applications/grovers.py

```diff
@@ -0,0 +1,244 @@
+
+    q_c.h(num_qubits - 1)
+
+    # q_c.barrier()
+
+    for (q_var, bit) in enumerate(marked_item_bits):
+        if not int(bit):
+            q_c.x(q_var)
+
+    if grover_oracle is None or num_qubits <= 5:
+        if num_qubits < 9:
+            grover_oracle = q_c
+
+    return q_c
+
+
+############## Grover Diffusion Operator
+
+
+def add_diffusion_operator(num_qubits):
+    """This function creates the diffusion operator for grover's circuit"""
+    diffusion_operator = None
+    q_r = QuantumRegister(num_qubits)
+    q_c = QuantumCircuit(q_r, name="diffuser")
+
+    for i_qubit in range(num_qubits):
+        q_c.h(q_r[i_qubit])
+    for i_qubit in range(num_qubits):
+        q_c.x(q_r[i_qubit])
+    q_c.h(num_qubits - 1)
+
+    if _USE_MCX_SHIM:
+        add_mcx(q_c, list(range(num_qubits - 1)), num_qubits - 1)
+    else:
+        q_c.mcx(list(range(num_qubits - 1)), num_qubits - 1)
+
+    q_c.h(num_qubits - 1)
+
+    # q_c.barrier()
+
+    for i_qubit in range(num_qubits):
+        q_c.x(q_r[i_qubit])
+    for i_qubit in range(num_qubits):
+        q_c.h(q_r[i_qubit])
+
+    if diffusion_operator is None or num_qubits <= 5:
+        if num_qubits < 9:
+            diffusion_operator = q_c
+
+    return q_c
+
+
+############### MCX shim
+
+# single cx / cu1 unit for mcx implementation
+def add_cx_unit(q_c, cxcu1_unit, controls, target):
+    """This function creates the MCX shim for grover's circuit"""
+    num_controls = len(controls)
+    i_qubit = cxcu1_unit[1]
+    j_qubit = cxcu1_unit[0]
+    theta = cxcu1_unit[2]
+
+    if j_qubit is not None:
+        q_c.cx(controls[j_qubit], controls[i_qubit])
+    q_c.cu1(theta, controls[i_qubit], target)
+
+    i_qubit = i_qubit - 1
+    if j_qubit is None:
+        j_qubit = i_qubit + 1
+    else:
+        j_qubit = j_qubit - 1
+
+    if theta < 0:
+        theta = -theta
+
+    new_units = []
+    if i_qubit >= 0:
+        new_units += [[j_qubit, i_qubit, -theta]]
+        new_units += [[num_controls - 1, i_qubit, theta]]
+
+    return new_units
+
+
+# mcx recursion loop
+def add_cxcu1_units(q_c, cxcu1_units, controls, target):
+    """This function recursively creates the MCX shim for grover's circuit"""
+    new_units = []
+    for cxcu1_unit in cxcu1_units:
+        new_units += add_cx_unit(q_c, cxcu1_unit, controls, target)
+    cxcu1_units.clear()
+    return new_units
+
+
+def add_mcx(q_c, controls, target):
+    """mcx gate implementation: brute force and inefficent
+    start with a single CU1 on last control and target
+    and recursively expand for each additional control"""
+    num_controls = len(controls)
+    theta = np.pi / 2**num_controls
+    q_c.h(target)
+    cxcu1_units = [[None, num_controls - 1, theta]]
+    while len(cxcu1_units) > 0:
+        cxcu1_units += add_cxcu1_units(q_c, cxcu1_units, controls, target)
+    q_c.h(target)
+
+
+# def grovers_search(num_qubits, marked_item, n_iterations)
+def grovers_dist(num_qubits, marked_item, n_iterations):
```

Comment:
What is this function doing exactly? It doesn't seem to be called anywhere else in the file. Is it needed for the benchmarking?

Reply:
This code just creates a dictionary that can be used to visualize the amplitudes of each state after the circuit. But, it isn't necessary for the benchmark, so I'll remove.